### PR TITLE
samd21:setup all clock generators in clock.c

### PR DESCRIPTION
**Pull request by photonthunder**
Goal is to have all clock generators setup in clock.c and then turned on/off in periph_conf.c.  Does contain some cleanup as well.

### CODE REVIEW

File: boards/samd21-xpro/include/periph_conf.h [hunk 1]
```diff
@@ -272,6 +277,12 @@ static const spi_conf_t spi_config[] = {
 #define RTT_MAX_VALUE       (0xffffffff)
 #define RTT_FREQUENCY       (32768U)    /* in Hz. For changes see `rtt.c` */
 #define RTT_RUNSTDBY        (1)         /* Keep RTT running in sleep states */
+
+#if RTT_NUMOF || RTC_NUMOF
+#ifndef XOSC32_RUNSTBY
```
Comment (kaspar030) [inline boards/samd21-xpro/include/periph_conf.h]: seems like a D is missing

File: boards/samd21-xpro/include/periph_conf.h [hunk 2]
```diff
@@ -73,16 +75,19 @@ extern "C" {
...
-#define CLOCK_8MHZ          (1)
-#define GEN2_ULP32K         (1)
-#else
+#define CLOCK_CORECLOCK     (48000000UL)
+#elif CLOCK_USE_8MHZ_DEFAULT
 /* edit this value to your needs */
 #define CLOCK_DIV           (1U)
 /* generate the actual core clock frequency */
 #define CLOCK_CORECLOCK     (8000000 / CLOCK_DIV)
 #endif
```
Comment (kaspar030) [inline boards/samd21-xpro/include/periph_conf.h]: Does it make sense to add an error case here?

**Comment by dylad** [comment]
I'll try to take a deeper look soon but what I can already said : there are some broken boards with the current changes 
- sodaq autonomo
- arduino zero
- arduino mkr1000

File: boards/samd21-xpro/include/periph_conf.h (hunk 2 above)
Comment (photonthunder) [inline boards/samd21-xpro/include/periph_conf.h]: Added, I wish there was a better way to set this up to make it cleaner...

File: boards/samd21-xpro/include/periph_conf.h (hunk 1 above)
Comment (photonthunder) [inline boards/samd21-xpro/include/periph_conf.h]: Nice catch

**Comment by biboc** [comment]
That's a good idea to move everything in clock.c
Do it for SAML21 as well, it is very similar. 
Should you move clock.c in sam0_common folder?

**Comment by photonthunder** [comment]
@biboc, @dylad had made the same request.  I was thinking we would get this approved and then expand to the SAML21, but we can expand it now to help get it moving.  Mostly waiting for feedback on where to put the `#ifndef` so that you don't have to include all the options in every periph_conf.h.

**Comment by photonthunder** [comment]
Any feedback on these generic clock generators?  Need to clean this up since both RTT and the internal 32 kHz oscillator try to use generator 2 you can end up with a very difficult bug to find.  Almost need a periph_clock.h to go along with periph_conf.h.  You put the default settings (#ifndef) in periph_clock.h and place that after periph_conf.h in the appropriate files.  I will experiment but would greatly appreciate some ideas.

**Comment by photonthunder** [comment]
@biboc, @dylad I was just looking at the saml21 board and I don't see any code referring to the different generic clock generators?  Don't even setup the external oscillator for rtt.c?  So I think I will try and get this working for the samd21 only for now.

**Comment by photonthunder** [comment]
@kaspar030 I moved the ifndef to an include file (periph_clock_config.h) and added it to the appropriate samd21 files.  All boards now compile.  Rebased and Squashed as well.

**Comment by dylad** [comment]
>  So I think I will try and get this working for the samd21 only for now

Ok, once your PR will be merged, I'll expand your work to SAML21 family.

File: boards/arduino-mkr1000/include/periph_conf.h [hunk 3]
```diff
@@ -34,16 +34,17 @@ extern "C" {
 #endif
 
 /**
- * @brief   External oscillator and clock configuration
+ * @name   External oscillator and clock configuration
```
Comment (haukepetersen) [inline boards/arduino-mkr1000/include/periph_conf.h]: space missing

File: boards/arduino-mkr1000/include/periph_conf.h [hunk 4]
```diff
@@ -64,19 +65,28 @@ extern "C" {
...
  * @{
  */
-#define CLOCK_USE_PLL       (1)
+#define CLOCK_USE_PLL               (1)
+#define CLOCK_USE_XOSC32_DFLL       (0)
+#define CLOCK_USE_8MHZ_DEFAULT      (0)
 
 #if CLOCK_USE_PLL
-/* edit these values to adjust the PLL output frequency */
+    /* edit these values to adjust the PLL output frequency */
```
Comment (haukepetersen) [inline boards/arduino-mkr1000/include/periph_conf.h]: these indention seems to be off

File: cpu/samd21/include/periph_clock_config.h [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+#ifndef PERIPH_CLOCK_CONFIG_H
+#define PERIPH_CLOCK_CONFIG_H
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief   Default Clock configurations
+ */
```
Comment (haukepetersen) [inline cpu/samd21/include/periph_clock_config.h]: I think doxygen will be more happy if you use
```c
/**
 * @name ...
 * @{
 */

/** @} */
```
here.

File: cpu/samd21/include/periph_clock_config.h [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    
+/* If RTC or RTT is on Need XOSC32 */
```
Comment (haukepetersen) [inline cpu/samd21/include/periph_clock_config.h]: is **in** need **of**?

File: cpu/samd21/include/periph_clock_config.h [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+#if RTC_NUMOF || RTT_NUMOF
+#ifndef CLOCK_XOSC32K
+#define CLOCK_XOSC32K       (32768UL)
+#endif
+#ifndef GEN2_XOSC32
+#define GEN2_XOSC32         (1)
+#endif
+#endif
```
Comment (haukepetersen) [inline cpu/samd21/include/periph_clock_config.h]: I must say I am not so happy with the RTC/RTT dependency here. I would prefer that the cpu clock module merely provides functions to enable/disable the low frequency clock, which then in term can be used by the RTT/RTC. This is how we recently refactored the clock code for the stm32 family...

Comment (haukepetersen) [reply cpu/samd21/include/periph_clock_config.h]: So more precise: IMHO the board config should only define if an external 32khz oscillator is available, and the RTC/RTT modules should then simply request enabling that clock from this module. 

Or of course the XOSC32K clock is enabled during system initialization in case it is configured to be feeding the PLL.

**Comment by photonthunder** [comment]
@haukepetersen thanks for looking at this.  I do like your comments about core clock configuration since I imagine more options will be introduce in the future.  I also like the idea of doing an enable/disable function, I will look at the stm32 code and try to mimic.

**Comment by photonthunder** [comment]
In this version there is just two parameters visible by default for core clock configuration (CLOCK_SRC and CLOCK_CORECLOCK).  Interesting thing is that I found what could be a bug in gpio.c where it uses the generic clock generator 2.  Well in the old code clock generator 2 was either the ultra low power 32 kHz or the external oscillator.  All three modes do appear to work for the samd21-xpro.

**Comment by travisgriggs** [comment]
Hi @kaspar030 and @haukepetersen. Are we on the right direction with this still here? We would love to see this go in. It would make our own maintenance with RIOT a lot simpler since we use this in production.

**Comment by photonthunder** [comment]
updated external interrupt gpio so that it uses XOSC32 if it is available.  Gives better response on interrupt.

**Comment by photonthunder** [comment]
@kaspar030, @haukepetersen can I get a review on this one?

File: cpu/samd21/cpu.c [hunk 8]
```diff
@@ -15,41 +15,151 @@
...
+#include "periph_clock_config.h"
 #include "periph/init.h"
 
-#ifndef CLOCK_8MHZ
-#define CLOCK_8MHZ          1
+/**
+ * @name   Default Clock configurations
+ * @{
+ */
+
```
Comment (aabadie) [inline cpu/samd21/cpu.c]: This line can be removed

File: cpu/samd21/cpu.c [hunk 9]
```diff
@@ -15,41 +15,151 @@
...
-#define GEN2_ULP32K         1
+/* If never set then use these defaults */
+#ifndef CLOCK_XOSC32K
+#define CLOCK_XOSC32K           (0)
 #endif
 
 #ifndef VDD
 /**
- * @brief   Set system voltage level in mV (determines flash wait states)
+ * @brief set system voltage level in mV (determines flash wait states)
```
Comment (aabadie) [inline cpu/samd21/cpu.c]: Unrelated change

File: cpu/samd21/cpu.c [hunk 10]
```diff
@@ -15,41 +15,151 @@
...
  *
- * @note    Override this value in your boards periph_conf.h file
- *          if a different system voltage is used.
+ * @note  override this value in your boards periph_conf.h file
```
Comment (aabadie) [inline cpu/samd21/cpu.c]: Unrelated change

File: cpu/samd21/cpu.c [hunk 11]
```diff
@@ -155,7 +228,7 @@ static void clk_init(void)
                            SYSCTRL_DFLLMUL_MUL(CLOCK_CORECLOCK / CLOCK_XOSC32K);
     SYSCTRL->DFLLVAL.reg = SYSCTRL_DFLLVAL_COARSE(coarse) |
                            SYSCTRL_DFLLVAL_FINE(fine);
-    SYSCTRL->DFLLCTRL.reg = SYSCTRL_DFLLCTRL_MODE;
+                           SYSCTRL->DFLLCTRL.reg = SYSCTRL_DFLLCTRL_MODE;
```
Comment (aabadie) [inline cpu/samd21/cpu.c]: unrelated change

File: cpu/samd21/cpu.c [hunk 12]
```diff
@@ -15,41 +15,151 @@
...
+                         GCLK_GENCTRL_RUNSTDBY |
+                         GCLK_GENCTRL_DIVSEL |
+                         GCLK_GENCTRL_ID(3));
+
+    while (GCLK->STATUS.reg & GCLK_STATUS_SYNCBUSY) {}
+    is_enabled_gen3_ULP32k = true;
+}
+
+/**
+ * @brief configure clock sources and the cpu frequency
```
Comment (aabadie) [inline cpu/samd21/cpu.c]: You don't need doxygen in .c files, a simple comment is ok

File: cpu/samd21/cpu.c [hunk 13]
```diff
@@ -168,42 +241,41 @@ static void clk_init(void)
...
     /* redirect all peripherals to a disabled clock generator (7) by default */
     for (int i = 0x3; i <= 0x22; i++) {
-        GCLK->CLKCTRL.reg = ( GCLK_CLKCTRL_ID(i) | GCLK_CLKCTRL_GEN_GCLK7 );
-        while (GCLK->STATUS.bit.SYNCBUSY) {}
+        GCLK->CLKCTRL.reg = (GCLK_CLKCTRL_ID(i) |
+                             GCLK_CLKCTRL_GEN_GCLK7);
+        while (GCLK->STATUS.reg & GCLK_STATUS_SYNCBUSY) {}
     }
 }
+/** @} */
```
Comment (aabadie) [inline cpu/samd21/cpu.c]: can be removed

File: cpu/samd21/cpu.c [hunk 14]
```diff
@@ -15,41 +15,151 @@
...
 
 #include "cpu.h"
 #include "periph_conf.h"
+#include "periph_clock_config.h"
 #include "periph/init.h"
 
-#ifndef CLOCK_8MHZ
-#define CLOCK_8MHZ          1
+/**
+ * @name   Default Clock configurations
```
Comment (aabadie) [inline cpu/samd21/cpu.c]: You don't need doxygen in .c files, simple comments are ok

File: cpu/samd21/include/periph_clock_config.h [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+
+#ifndef PERIPH_CLOCK_CONFIG_H
+#define PERIPH_CLOCK_CONFIG_H
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @name generic clock generator 1
```
Comment (aabadie) [inline cpu/samd21/include/periph_clock_config.h]: Only use `@brief` and no opening/closing block (`@{ @}`)

File: cpu/samd21/include/periph_cpu.h [hunk 16]
```diff
@@ -29,6 +29,13 @@ extern "C" {
 #endif
 
 /**
+ * @brief   Available core clock configurations
```
Comment (aabadie) [inline cpu/samd21/include/periph_cpu.h]: Here you can use `@name` with opening/closing block before and after the 3 defines.

File: cpu/samd21/cpu.c [hunk 17]
```diff
@@ -15,41 +15,151 @@
...
 #else
-#define WAITSTATES          ((CLOCK_CORECLOCK - 1) / 14000000)
+#define WAITSTATES              ((CLOCK_CORECLOCK - 1) / 14000000)
 #endif
 
 /**
- * @brief   Configure clock sources and the cpu frequency
+ * @brief setup 1 MHz (8MHz clock/8) clock on Generic Clock Generator 1
+ */
+static bool is_enabled_gen1_1MHz = false;
```
Comment (aabadie) [inline cpu/samd21/cpu.c]: I'm unsure if this static variable is really required ? Why do you want to prevent multiple setups ?

**Comment by photonthunder** [comment]
@aabadie thanks for the review.  I obviously struggle with the doxygen labeling so thanks for the insight.  I have the checks, such as is_enabled_gen1_1MHz, because I have seen cases where modifying  while it is on and being used can cause issues without first turning it off (like what you see in xosc32).

File: cpu/samd21/include/periph_clock_config.h [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief generic clock generator 1
+ * setup 1 MHz clock on generic clock generator 1
+ */
+uint8_t setup_gen1_1MHz(void);
```
Comment (aabadie) [inline cpu/samd21/include/periph_clock_config.h]: AFAICS, the returned value is not taken into account where this function is called, maybe use `void` instead of `uint8_t`

File: cpu/samd21/include/periph_clock_config.h [hunk 19]
```diff
@@ -0,0 +1,49 @@
...
+
+/**
+ * @brief generic clock generator 2
+ * setup 32 kHz XOSC on generic clock generator 2
+ */
+void setup_gen2_xosc32(bool standby);
```
Comment (aabadie) [inline cpu/samd21/include/periph_clock_config.h]: the standby parameter is not documented (static tests will complain about that...)

**Comment by photonthunder** [comment]
Had to rebase and resolve conflicts with changes in the latest master so I squashed as well...

**Comment by photonthunder** [comment]
Any chance I can get some more feedback on this one?

**Comment by photonthunder** [comment]
So I am curious if the changes in here are too much for people to want to push along.  Should I make an attempt to push these changes in separate PRs that might be more palatable?

**Comment by dylad** [comment]
I don't think there is a need to split this PR.
Everyone has been very busy lately this is why your PR is not merged yet.

**Comment by photonthunder** [comment]
@dylad thanks for the feedback.  I will just keep it rebased and wait.

**Comment by dylad** [comment]
`sodaq-explorer` board is missing from the list.

**Comment by photonthunder** [comment]
Thanks @dylad, I will add it.  Is there a command that allows to to check all boards that are used by a certain cpu.  Something like make CPU=samd21?  I think I saw posts for the feather board which would also use the samd21.

**Comment by dylad** [comment]
I'm sure older RIOT's users have a better technique but personally I used
`grep -r -l "export CPU = samd21"`
or
`grep -r -l "export CPU = saml21"`

**Comment by photonthunder** [comment]
@dylad thanks for the info.

**Comment by photonthunder** [comment]
@aabadie thanks, looks like I need to add the feather as well...

**Comment by photonthunder** [comment]
Added both the feather-m0 and the soda-explorer

**Comment by aabadie** [comment]
I just this PR a try on samr21-xpro using the `tests/periph_timer` application and it failed:
```
$ make BOARD=samr21-xpro-C tests/periph_timer/ flash term

[...]

2017-11-14 09:39:49,783 - INFO # Available timers: 2
2017-11-14 09:39:49,783 - INFO # 
2017-11-14 09:39:49,784 - INFO # Testing TIMER_0:
2017-11-14 09:39:49,787 - INFO # TIMER_0: initialization successful
2017-11-14 09:39:49,789 - INFO # TIMER_0: stopped
2017-11-14 09:39:49,792 - INFO # TIMER_0: ERROR setting any channel
2017-11-14 09:39:49,792 - INFO # 
2017-11-14 09:39:49,792 - INFO # 
2017-11-14 09:39:49,794 - INFO # Testing TIMER_1:
2017-11-14 09:39:49,797 - INFO # TIMER_1: initialization successful
2017-11-14 09:39:49,798 - INFO # TIMER_1: stopped
2017-11-14 09:39:49,801 - INFO # TIMER_1: ERROR setting any channel
2017-11-14 09:39:49,802 - INFO # 
2017-11-14 09:39:49,802 - INFO # 
2017-11-14 09:39:49,803 - INFO # TEST FAILED
```

I get the same result on `arduino-zero`.

**Comment by photonthunder** [comment]
Thanks for checking, I use this code in my build but we don’t currently use the RIOT timer code.  Probably a recent branch messed it up. I will fix it...

**Comment by photonthunder** [comment]
@aabadie, I had missed a merge fix because of some space issues.  Cleaned up and is now working

**Comment by aabadie** [comment]
Thanks @photonthunder, will give a try again tomorrow

File: cpu/samd21/cpu.c [hunk 20]
```diff
@@ -15,42 +15,136 @@
...
+        gen2_enabled = false;
+        run_standby_gen2_xosc32 = true;
+    }
+    if (gen2_enabled) {
+        return;
+    }
+
+    /* Use External 32.768KHz Oscillator */
+    SYSCTRL->XOSC32K.reg = SYSCTRL_XOSC32K_ONDEMAND |
+    SYSCTRL_XOSC32K_EN32K |
```
[the remark this answers is not in the document]
Comment (photonthunder) [reply cpu/samd21/cpu.c]: Thanks for catching this, I thought I had cleaned them all up but I obviously missed this one.

**Comment by photonthunder** [comment]
@aabadie, I am not seeing any issues when I switch to the 8 MHz option:

```
#define CLOCK_SRC                   (CLOCK_USE_8MHZ)
#define CLOCK_CORECLOCK             (8000000U)
```

Was there a particular test that wasn't showing output?

**Comment by aabadie** [comment]
> #define CLOCK_SRC                   (CLOCK_USE_8MHZ)
> #define CLOCK_CORECLOCK             (8000000U)

> Was there a particular test that wasn't showing output?

Hmm, so the coreclock has to be changed as well... Maybe it can be set automatically when `CLOCK_SRC` is defined to `CLOCK_USE_8MHZ` ?

I think I was tested the `tests/periph_timer` application. Will try again tomorrow.

**Comment by photonthunder** [comment]
@aabadie the PLL version allows for variation of CLOCK_CORECLOCK so I would have to have some if, else if similar to the way it used to be.  I was trying to real minimize the options for this version.  That being said I realize there needs to be more configuration options and thus I started issue #7913.  For this version, I am happy to add some configuration options back in if that is your preference?

**Comment by madokapeng** [comment]
Hello, I am new to RIOT project. where we are at this issue? Please be advised. @photonthunder

**Comment by dylad** [comment]
I think we can leave this PR as it for now and just add a small notice for users to warn them to update the CORECLOCK.

**Comment by tcschmidt** [comment]
@haukepetersen can we re-shift this into gear?

**Comment by dylad** [comment]
I guess all comments have been adressed so if @photonthunder is willing to rebase/squash it so we can finally move forward.

**Comment by Hyungsin** [comment]
Cool. Looking forward to it.

**Comment by photonthunder** [comment]
rebased and squashed.  But it has been a year since I touched this code...

**Comment by Hyungsin** [comment]
@photonthunder, please fix white space errors.

**Comment by photonthunder** [comment]
Sorry, was on vacation.  White space removed and squashed/rebased again.

**Comment by dylad** [comment]
@photonthunder looks like Travis is complaining again. Still some whitespaces in the wild.

**Comment by dylad** [comment]
Murdock approves. I'll try the PR on a samd21 ASAP just to make sure boards still run.

**Comment by photonthunder** [comment]
@dylad, I know this is going to sound crazy but I think that is correct since the default option is running  without standby on, so RTC doesn't run in deep sleep mode.  This is why I started Issue #7913 to try and come up with a better approach.  Will you try the same test but change `setup_gen2_xosc32(false);` to true and check the result?

**Comment by dylad** [comment]
THB, I don't think its a good idea to disable RTC by default. Users may want to use calendar, or trigger an event at specific time.
As long as MODULE_PERIPH_RTC is present, RTC must be enable for user.

**Comment by photonthunder** [comment]
So the existing code for RTC has standby off by default.  That doesn't mean RTC is off it just means that it doesn't run in sleep.  I think this is the larger issue that needs to be addressed with these chips is a method for indicating that whether or not you want things running in standby mode, which needs to occur when you setup the initial clock generators.  Thus you need a whole bunch of defines which quickly gets overwhelming for users.

**Comment by dylad** [comment]
@photonthunder I agree with you. We need a way to enable/disable standby mode efficiently. But in the current state, tests/periph_rtc doesn't work with this PR (standby mode or not) this need to be fix. We can then think about how to enable/disable standby features later.

**Comment by photonthunder** [comment]
@dylad sounds good, so did you try it with with standby enabled?  setup_gen2_xosc32(true); Did the test pass?

**Comment by dylad** [comment]
I cannot run any tests as I am on vacation right now.
If someone can test it would be great.
Anyway, If this works  with setup_gen2_xosc32(true) then this PR need to be adapted as well

**Comment by photonthunder** [comment]
Updated to true...

**Comment by dylad** [comment]
I just retested this PR, (Sorry for the delay, I had to order another board because the previous died)
tests/periph_rtc still don't have the required behaviour.

**Comment by keestux** [comment]
What's the status of this PR?

Oh, and by the way, the commit message has "...all clock generators in clock.c", but there is no clock.c. Is that an old commit message?

**Comment by photonthunder** [comment]
The whole thing is very old.  I explored other methods to try and handle the challenge that each peripheral is tied back to a GCLK and how to make that interactive without so many options (see #7913) but I haven't touched this in a long time so it probably needs a clean look..  probably just close this and someone can start a new PR if they desire...

**Comment by keestux** [comment]
Testing with Sodaq ONE (with a modified periph_conf.h, because it is missing in the patch) gives incorrect timing of `tests/periph_rtc`. All Alarm messages are printed at once. There should be a 2s delay between them.

`tests/periph_rtt` seems to behave properly.

**Comment by photonthunder** [comment]
Feel free to hack away if you have time and interest.  I may be able to look at this later this week or next week.  I just figured most people where just hacking what they needed for each specific board?

**Comment by keestux** [comment]
I should have shown more interest long time ago. Sorry about that.
